### DISCUSSION TITLE
Notebook 02: Graphs Turning Orange

### DISCUSSION THREAD

**Discussion by apyhyk** [mrdbourke/pytorch-deep-learning#467] [Q&A]
I am on Chapter 2 – Neural Network Classification and whenever I try and run the graph for anything but model 0 the graph comes out orange, I have no idea what is happening or why

here is my code: 

## Make and plot data 
```
import torch
import torch.nn as nn
import torch.optim as optim
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
from sklearn.datasets import make_circles


n_samples = 1000

X, y = make_circles(n_samples,
                    noise=0.03,
                    random_state=42)
plt.scatter(X[:, 0],X[:, 1], c=y, cmap=plt.cm.RdBu)`

```
## Turn data into tensors
```
X = torch.from_numpy(X).type(torch.float)
y = torch.from_numpy(y).type(torch.float)
```

## Split into train and test sets
```
X_train, X_test, y_train, y_test = train_test_split(X,
                                                    y,
                                                    test_size=0.2,
                                                    random_state=42)
X_train[:5], y_train[:5]
```

## Build a model with non-linear activation functions 
```
from torch import nn
class CircleModelV2 (nn.Module):
  def __init__(self):
    super().__init__()
    self.layer_1 = nn.Linear(2,10)
    self.layer_2 = nn.Linear(10,10)
    self.layer_3 = nn.Linear(10,1)
    self.relu = nn.ReLU()
    
  def forward(self, x): # relu is a non-linear activation function
    # Where should we put our non-linear activation function
    return self.layer_3(self.relu(self.layer_2(self.relu(self.layer_1(x)))))
  
model_3 = CircleModelV2().to(device)
model_3.to(device)


```

## Loss // Optimizer: 
```

loss_fn = nn.BCEWithLogitsLoss()
optimizer = torch.optim.SGD(model_3.parameters(), lr=0.01)
```

## To Device: 

```
X_train, y_train = X_train.to(device), y_train.to(device)
X_test, y_test = X_test.to(device), y_test.to(device)
```

## Building and testing a model:
```

epochs = 1000

for epoch in range(epochs):
  model_3.train()
  y_logits = model_3(X_train).squeeze()
  y_pred = torch.round(torch.sigmoid(y_logits))
  loss = loss_fn(y_logits, y_pred) ## RAW DATA COMPARED WITH PREDICTED DATA 
  acc = accuracy_fn(y_true=y_train,
                    y_pred=y_pred)
  optimizer.zero_grad()
  loss.backward()
  optimizer.step()

  model_3.eval()
  with torch.inference_mode():
    test_logits = model_3(X_test).squeeze()
    test_pred = torch.round(torch.sigmoid(test_logits))
    test_loss = loss_fn(test_logits, test_pred)
    test_acc = accuracy_fn(y_true=y_test,
                           y_pred=test_pred)

  if epoch % 100 == 0:
        print(f"Epoch: {epoch} | Loss: {test_loss: .5f}, | Test Loss: {test_loss: .5f}")

```

## Makes predictions
```
model_3.eval()
with torch.inference_mode():
  y_preds = torch.round(torch.sigmoid(model_3(X_test))).squeeze()
y_preds[:10], y_test[:10]
```

## Plot decision boundaries
```
plt.figure(figsize=(12, 6))
plt.subplot(1, 2, 1)
plt.title("Train")
plot_decision_boundary(model_3, X_train, y_train) # model_1 = no non-linearity
plt.subplot(1, 2, 2)
plt.title("Test")
plot_decision_boundary(model_3, X_test, y_test) # model_3 = has non-linearity
```

![image](https://github.com/mrdbourke/pytorch-deep-learning/assets/118761234/e4f713a8-c009-4f3b-bb0b-113ed82d3bec)

If I do this with model_0 here is the result: 

```
# Plot decision boundaries
plt.figure(figsize=(12, 6))
plt.subplot(1, 2, 1)
plt.title("Train")
plot_decision_boundary(model_0, X_train, y_train) # model_1 = no non-linearity
plt.subplot(1, 2, 2)
plt.title("Test")
plot_decision_boundary(model_0, X_test, y_test) # model_3 = has non-linearity
```

![image](https://github.com/mrdbourke/pytorch-deep-learning/assets/118761234/56e9dc89-1eeb-4f83-85df-5e37f48988d1)

model_1 here is the result: 

```
# Plot decision boundaries
plt.figure(figsize=(12, 6))
plt.subplot(1, 2, 1)
plt.title("Train")
plot_decision_boundary(model_1, X_train, y_train) # model_1 = no non-linearity
plt.subplot(1, 2, 2)
plt.title("Test")
plot_decision_boundary(model_1, X_test, y_test) # model_3 = has non-linearity
```

![image](https://github.com/mrdbourke/pytorch-deep-learning/assets/118761234/fcdc339e-9518-4f07-ac64-2222ee7aca36)

# ORIGINAL HELPER FUNCTION CODE: 
```

import requests
from pathlib import Path

# Download helper functions from Learn PyTorch repo (if it's not already downloaded)
if Path("helper_functions.py").is_file():
  print("helper_functions.py already exists, skipping download")
else:
  print("Downloading helper_functions.py")
  request = requests.get("https://raw.githubusercontent.com/mrdbourke/pytorch-deep-learning/main/helper_functions.py")
  with open("helper_functions.py", "wb") as f:
    f.write(request.content)

from helper_functions import plot_predictions, plot_decision_boundary
# Plot decision boundary of the model
plt.figure(figsize=(12, 6))
plt.subplot(1, 2, 1)
plt.title("Train")
plt.subplot(1, 2, 2)
plt.title("Test")

```

**Comment by mrdbourke** [comment]
Hi @apyhyk ,

I've just run the original [notebook 02](https://www.learnpytorch.io/02_pytorch_classification/) end-to-end and found the code to be performing as expected.

I've found two differences in your code to the original notebook 02 so far: 

1. Learning rate for `model_3` (yours = `0.01` and notebook 02 = `0.1`)
2. Loss calculation for `model_3` (yours = compute on `y_pred` and notebook 02 = compute on `y_train`)

## Learning rate

Your code:

```python
loss_fn = nn.BCEWithLogitsLoss()
optimizer = torch.optim.SGD(model_3.parameters(), lr=0.01)
```

Notebook 02 code:

```python
# Setup loss and optimizer 
loss_fn = nn.BCEWithLogitsLoss()
optimizer = torch.optim.SGD(model_3.parameters(), lr=0.1)
```

## Loss calculation

Notice the difference with computing the loss.

Your code: 

```python
for epoch in range(epochs):
  model_3.train()
  y_logits = model_3(X_train).squeeze()
  y_pred = torch.round(torch.sigmoid(y_logits))
  loss = loss_fn(y_logits, y_pred) ## RAW DATA COMPARED WITH PREDICTED DATA 
```

Notebook 02 code:

```python
for epoch in range(epochs):
    # 1. Forward pass
    y_logits = model_3(X_train).squeeze()
    y_pred = torch.round(torch.sigmoid(y_logits)) # logits -> prediction probabilities -> prediction labels
    
    # 2. Calculate loss and accuracy
    loss = loss_fn(y_logits, y_train) #### Note: y_logits vs y_train
```

Try these updates and see how you go.

Otherwise, the original notebook (no errors) I just ran can be seen here: https://colab.research.google.com/drive/14kQao_zmh0wNrh-V1jbky4dSlAJsMkNg?usp=sharing
